[IMP] website: replace popup by popover on menu edition in edit mode

rdeodoo opened the pull request:
Before this commit, in edit mode, when clicking on a menu, a popup would be
shown to ask the user what he wants to do (edit the menu or go to the link or
do nothing). That popup was a bit invasive, overkill and old-fashioned.

Now, a simple & light popover/tooltip will be shown instead with the same
options as in Google Doc when editing a link.

This commit also takes the opportunity to finally let regular dropdown menus
open in edit mode so user can edit it. This was mandatory for the popover
behavior on those menus.

task-2439860

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/64756).

rdeodoo commented:
Result: https://drive.google.com/file/d/1GIripGSy_ouOlKQ36kKG9ZhhsmfYELSe/view

rdeodoo commented:
New code to works with dropdown menu too.
Removed the hackish code needed to have a custom behavior for popover (open on click & close on mouseout), instead just used regular click behavior.
Chance was taken to also let dropdown open in edit mode (which was needed for this PR feature but also wanted previously)

JKE-be commented:
ca semble deja plus clean ainsi ;)

JKE-be commented on addons/website/controllers/main.py:
```diff
@@ -291,6 +292,45 @@ def get_dynamic_snippet_templates(self, filter_id=False):
         return templates
 
+    @http.route('/website/get_meta_from_url', type='json', auth='user', website=True)
+    def get_meta_from_url(self, url):
```
on avait pas dis client side ?

JKE-be replied:
and just favicon / title

rdeodoo replied:
Cette route n'est pas utilisée c'est du old code de l'implémentation précédente, je le garde sous la main tant que la PR n'est pas ready, comme le commit 'wip link in page' qui n'est plus pertinent.

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/link_popover_widget.js:
```diff
@@ -0,0 +1,182 @@
+     * @private
+     */
+    _resetPreview() {
+        this.$('.js_preview_favicon img').addClass('d-none');
+        this.$('.js_preview_favicon .fa').removeClass('d-none');
+        this.$('.js_preview_description').addClass('d-none');
+        this.$('.js_preview_image').addClass('d-none');
+        this.$('.js_preview_site').text("");
+        // Useful for 404 internal URL
+        this.$('.js_url_link').text(this.$target.attr('href'));
```
Since you search for those here and also during link update & stuff, I would save the search in a variable and use that variable. That would avoid repeating the selectors.

qsm-odoo commented on addons/website/static/src/js/widgets/link_popover_widget.js:
```diff
@@ -0,0 +1,97 @@
+     * @param {Event} ev
+     */
+     _onEditMenuClick(ev) {
+        const self = this;
```
No self :)

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1067,45 +1068,28 @@ options.registry.ThemeColors = options.registry.OptionsTab.extend({
-        })).open();
+    start: function () {
+        this.LinkPopoverWidget = new wLinkPopoverWidget(this, this.$target[0]);
+        this.LinkPopoverWidget.appendTo(this.$target[0]);
```
Detail but `appendTo` returns a promise, which you could `await` here.

qsm-odoo commented on addons/website/static/src/js/widgets/link_popover_widget.js:
```diff
@@ -0,0 +1,97 @@
+odoo.define('website.widgets.LinkPopoverWidget', function (require) {
```
Odoo module here too

qsm-odoo reviewed (changes requested):
Superficial review then we can send to testing, but I don't think there is anything else to say anyway :) I tested the branch a bit and saw some things to fix / improve:

- 
![Screenshot from 2021-03-04 18-05-04](https://user-images.githubusercontent.com/10338094/110001429-a9694b00-7d14-11eb-93ba-ffcddbd2f4e9.png)

- Break link feature does not work (not related to your branch though I think... but cannot remove style of buttons

- I would not close the popover when clicking inside of it but not on a button... possible?

- I would squash the last 3 commits ;)

rdeodoo commented:
Everything has been processed.

> ![Screenshot from 2021-03-04 18-05-04](https://user-images.githubusercontent.com/10338094/110001429-a9694b00-7d14-11eb-93ba-ffcddbd2f4e9.png)

Fixed, indeed that value was only set from the template during rendering (widget.href).
> * Break link feature does not work (not related to your branch though I think... but cannot remove style of buttons

I have no issue on my end :/ see https://drive.google.com/file/d/1rc_BfS2orfdkJlcSpnOO6hLLaYRJfNjI/view

> * I would not close the popover when clicking inside of it but not on a button... possible?

[Of course I can, I can do anything, I can do absolutely anything... I am an expert.](https://youtu.be/BKorP55Aqvg?t=427)
I tried a bit with the first implementation and it was a bit annoying to do given the place where the code was initiated in the editor, but I'll do it now.

> * I would squash the last 3 commits ;)

Done

rdeodoo commented:
Also, I will see how to show the URL inside the popover without noising it too much as it is an useful info to have directly.
cc @JKE-be

rdeodoo commented:
As mentioned by @JKE-be , showing the full URL inside the popover is an useful information to have.
At first I wasn't in favor of it but after discussing about it, it is useful to directly see url params such as `utm` etc.
At the same time I don't want the popover to be 'noised' with (very) (long) urls -> if they got params, they are probably quite long.
This is what I suggest for now:
- Show the url with ellipsis, and on hover show the full URL (keep it mind the behavior is not only for menu but for all page links).
- https://drive.google.com/file/d/15MYM9cd5NtYvtoDvd3o31xzL6CU45mCD/view
- I kept it encapsulated in a wip commit: https://github.com/odoo/odoo/pull/64756/commits/22c081062e9df14925438d3de5579b788742b941

_________________

> I would not close the popover when clicking inside of it but not on a button... possible?

Done, a bit verbose but you can see the diff in this force push : https://github.com/odoo/odoo/compare/3ed713211c9c522f2a6efc2a1dceaf229acbd5c7..1760575397a738b81011ad75a9e816156c361c72

qsm-odoo commented:
@rdeodoo I like it. There is also this extra possibility to maybe make it even better: https://css-tricks.com/almanac/properties/l/line-clamp/ that would allow to have the URL on multiple lines before the ellipsis. I guess I may find an URL on 2 or 3 lines not ugly (and so in that case no need of the hover feature for this kind of URL that span on only 2 or 3 lines). (I have the feeling that the average-size links will span over 2 lines?).

JKE-be commented:
randomly edit link doesn't work ...
And bug of show/hidden (video 28s)
https://drive.google.com/file/d/1AprK2vy_bGxfhuW54ZMVpEwDKpuNHSu3/view

word-break: break-all; for url will be probably better (on hover if you want)

I still don't see the interest to have the company?website? name (2nd line) that is your current and so always the same...
And as suggest in a precedent review, width it too small to have 3 buttons side by side. 

:thinking:
<:trollface:> And now that we will have the new editor with link dialog merged in right column this week, do we need this task ? </:trollface:>

qsm-odoo commented:
> And now that we will have the new editor with link dialog merged in right column this week, do we need this task ? 

I don't think that changes anything, still needed to "start" edition :wink: But yes increasing the width of the popup seems good, JKE showed me last week :+1:

JKE-be commented:
![image](https://user-images.githubusercontent.com/4848797/110454552-aaf59300-80c7-11eb-871f-bf8ff85fa068.png)

for track

rdeodoo commented:
@qsm-odoo 
> There is also this extra possibility to maybe make it even better: https://css-tricks.com/almanac/properties/l/line-clamp/ 

:+1: 

@JKE-be 

> randomly edit link doesn't work ...
> And bug of show/hidden (video 28s)
> https://drive.google.com/file/d/1AprK2vy_bGxfhuW54ZMVpEwDKpuNHSu3/view

It seems like a bug related to last push about keeping the popover open if you click inside it but not a button (you can see in your video you clicked next to the button then clicked the button, I just reproduced it that way too). Weird, but I will fix that.

> word-break: break-all; for url will be probably better (on hover if you want)

Indeed much better :+1: 

> I still don't see the interest to have the company?website? name (2nd line) that is your current and so always the same...

It is the og:sitename but it is not much useful now that we excluded the external URL preview from s2s so this can be removed.

> <:trollface:> And now that we will have the new editor with link dialog merged in right column this week, do we need this task ? </:trollface:>

<:trollface:> :cry: took me 5 minutes to understand how to create a link as it was hidden in the bottom of the right panel and ctrl-k is gone :grimacing: </:trollface:>

rdeodoo commented:
So what do we do here?

qsm-odoo commented:
Had a quick look, added a commit, let's make this RD Validation Green and let's let SBU decide if we go through India or not :+1:

JKE-be commented:
Green color is better than red for a runbot... like purple is better than green for ...  ;)

some bug of positionning https://drive.google.com/file/d/1Q1kOjXUV2iCc8W6LhNQDRLnBmMaNZhuS/view

[break link](https://user-images.githubusercontent.com/4848797/111755642-9cfafb80-8899-11eb-810b-53f5074119c3.png) ? vs Remove link ? 

Double click on boutton should jsut open modal [without preview](https://user-images.githubusercontent.com/4848797/111755584-8e144900-8899-11eb-96ca-9ef1bfc91f35.png) imo

Link on img not supported ? https://drive.google.com/file/d/1T7BBeC2IIPtExF36Qy_am3HbqhnTdQLq/view if by design or tech, lgfm

[Favicon broken](https://user-images.githubusercontent.com/4848797/111756556-a2a51100-889a-11eb-8c55-014f0c7806b6.png) with phone number and email

Favicon of domain, or from page (could differ, so both is ok for me, just need to choose) E.g. of fail: [here](https://user-images.githubusercontent.com/4848797/111757234-658d4e80-889b-11eb-8abe-7aa2283f8258.png


Not related but to fix right after to have the best edit link feature of the world ;) ...

anchor is wrongly computed: https://drive.google.com/file/d/1-95tBWqOYhqMQQrGpaWQENoF7VlIRS5M/view on edit link 

Flickering with loading... of anchor!  Can't we put loading inside anchor select, or on the right side instead of below ?

Search page doesn't work anymore: https://drive.google.com/file/d/1fXCthOKCyp2Z7kajrY2LkRUpOxESUf-g/view



else LGTM, wait and see the feedback of @sbuhl to move forward this task

sbuhl commented:
`Remove Link` sounds better for lambda user imo.

rdeodoo commented:
Everything from @sbuhl & @pierre-pde has been processed, mainly:
- Removed double click (which was already removed but somehow lost during the dev)
- :fist_raised: replaced by `fa-sitemap`
- Preview image removed
- Tooltip adapted (Remove Link / Open in new tab)

Everything should also be good from @JKE-be review 
> Green color is better than red for a runbot... like purple is better than green for ... ;)

That was annoying.. It is a bug in standard, also in previous versions, not related to my PR, in blog module there is a missing `require` that somehow appear when we introduce a new JS odoo module (probably changing the file order)

> some bug of positionning https://drive.google.com/file/d/1Q1kOjXUV2iCc8W6LhNQDRLnBmMaNZhuS/view

Fixed, good catch, it was the same for link too close from the bottom of the page in the footer, it was correctly aligned but then when the preview was loaded, it would go off screen. Just needed to refresh the position after the preview if fetched.

> Double click on boutton should jsut open modal without preview imo

No more double click as mentioned above

> Link on img not supported ? https://drive.google.com/file/d/1T7BBeC2IIPtExF36Qy_am3HbqhnTdQLq/view if by design or tech, lgtm

It was a choice but image is now removed

> Favicon broken with phone number and email

Good catch, fixed it, I chose to show a `fa-phone` and `fa-envelope-o` for those. Adapted the code to not crash on invalid URL such as Odoo editor unsupported protocol eg `geo:`, see comment in code.

> Favicon of domain, or from page (could differ, so both is ok for me, just need to choose) E.g. of fail: [here](https://user-images.githubusercontent.com/4848797/111757234-658d4e80-889b-11eb-8abe-7aa2283f8258.png

This is because it is a 404 on facebook side and somehow google doesn't return a favicon, see:
404 url: https://www.google.com/s2/favicons?sz=64&domain_url=facebook.com/brol
good url: https://www.google.com/s2/favicons?sz=64&domain_url=facebook.com/romain/
I'd leave it like that, you might get a hint something is wrong with the URL that way.

> Not related but to fix right after to have the best edit link feature of the world ;) ...

I can work on it later, and I think we should support more than just `tel:` and `mailto:` as protocol, at least `geo:` and `ftp:` ?

robodoo commented:
Merge method set to rebase and fast-forward

robodoo commented:
Staging failed: ci/runbot on c1dc2390d6c9bb62874e70de49e1ea417ae09e68 (view more at http://runbot.odoo.com/runbot/build/6968178)

qsm-odoo reviewed (approved):
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward
